packages/compartment-mapper/src/zip/buffer-reader.js

```diff
@@ -0,0 +1,225 @@
+// @ts-check
+/* eslint no-bitwise: ["off"] */
+
+const q = JSON.stringify;
+
+// This is a mask of all the bits of the high 32 bits of a 64 bit integer that
+// indicate a number that cannot be stored in the 53 bits of integer precision
+// in a double precision floating point number.
+const bitsTooHigh = (-1 << (53 - 32)) >> 0;
+
+export class BufferReader {
+  /**
+   * @param {ArrayBuffer} data
+   */
+  constructor(data) {
+    this.data = new Uint8Array(data);
+    this.length = this.data.length;
+    this.index = 0;
+    this.offset = 0;
+  }
+
+  /**
+   * @param {number} index
+   * @return {boolean} whether the read head can move to the given absolute
+   * index.
+   */
+  canSeek(index) {
+    return index >= 0 && this.offset + index <= this.length;
+  }
+
+  /**
+   * @param {number} index the index to check.
+   * @throws {Error} an Error if the index is out of bounds.
+   */
+  assertCanSeek(index) {
+    if (!this.canSeek(index)) {
+      throw new Error(
+        `End of data reached (data length = ${this.length}, asked index ${index}`
+      );
+    }
+  }
+
+  /**
+   * @param {number} index
+   * @return {number} prior index
+   */
+  seek(index) {
+    const restore = this.index;
+    this.assertCanSeek(index);
+    this.index = index;
+    return restore;
+  }
+
+  /**
+   * @param {number} size
+   * @returns {Uint8Array}
+   */
+  peek(size) {
```

Comment:
Wanna guard against negative `size` here? Perhaps a job for `Nat`?

Reply:
I’ll roll it by hand for now. That seems most consistent with “WWWWWCD?” without crossing over to BigInt quite yet.